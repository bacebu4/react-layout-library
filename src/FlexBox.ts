import styled from "styled-components";

interface FlexBoxProps {
  jc?:
    | "space-around"
    | "space-between"
    | "space-evenly"
    | "stretch"
    | "center"
    | "end"
    | "flex-end"
    | "flex-start";
  ai?:
    | "center"
    | "end"
    | "flex-end"
    | "flex-start"
    | "self-end"
    | "self-start"
    | "start"
    | "baseline"
    | "normal"
    | "stretch";
  mt?: number;
  ml?: number;
  mr?: number;
  mb?: number;
  w?: string;
  h?: string;
  margin?: string;
  direction?: "column";
}

export const FlexBox = styled.div<FlexBoxProps>`
  display: flex;
  flex-direction: ${(props) => props.direction};
  justify-content: ${(props) => props.jc};
  align-items: ${(props) => props.ai};
  margin-left: ${(props) => props.ml}px;
  margin-right: ${(props) => props.ml}px;
  margin-bottom: ${(props) => props.ml}px;
  margin-top: ${(props) => props.mt}px;
  margin: ${(props) => props.margin};
  width: ${(props) => props.w};
  height: ${(props) => props.h};
`;
