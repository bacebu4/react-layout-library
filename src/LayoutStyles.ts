/* eslint-disable @typescript-eslint/ban-types */
import styled, { StyledComponent } from "styled-components";

type LayoutStylesProps = {
  mt?: number;
  ml?: number;
  mr?: number;
  mb?: number;
};

export function withLayoutStyles<P>(
  Component: React.FC<P>
): StyledComponent<React.FC<P>, object, P & LayoutStylesProps, never> {
  const ComponentWithAddedProps = styled(Component)<LayoutStylesProps & P>`
    ${(props) => {
      if (props.mt) {
        return `margin-top: ${props.mt}px;`;
      }
      return "";
    }}
    ${(props) => {
      if (props.mb) {
        return `margin-bottom: ${props.mb}px;`;
      }
      return "";
    }}
    ${(props) => {
      if (props.ml) {
        return `margin-left: ${props.ml}px;`;
      }
      return "";
    }}
    ${(props) => {
      if (props.mr) {
        return `margin-right: ${props.mr}px;`;
      }
      return "";
    }}
  `;

  return ComponentWithAddedProps;
}
