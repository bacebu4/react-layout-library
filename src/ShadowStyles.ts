/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/ban-types */
import styled, { StyledComponent } from "styled-components";

interface Options {
  x: number;
  y: number;
  b: number;
  s: number;
  color: string;
}

export function createShadowStyle(
  opts: Options = {
    x: 3,
    y: 2,
    b: 17,
    s: 2,
    color: "rgba(114, 114, 114, 0.15)",
  }
) {
  return function withCustomShadowStyles<P>(
    Component: React.FC<P>
  ): StyledComponent<React.FC<P>, any, {}, never> {
    const ComponentWithAddedProps = styled(Component)`
      box-shadow: ${opts.x}px ${opts.y}px ${opts.b}px ${opts.s}px ${opts.color};
    `;

    return ComponentWithAddedProps;
  };
}

export function withShadowStyles<P>(
  Component: React.FC<P>
): StyledComponent<React.FC<P>, any, {}, never> {
  const ComponentWithAddedProps = styled(Component)`
    box-shadow: 3px 2px 17px 2px rgba(114, 114, 114, 0.15);
  `;

  return ComponentWithAddedProps;
}
