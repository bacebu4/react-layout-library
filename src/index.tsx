import { withLayoutStyles } from "./LayoutStyles";
import { withShadowStyles, createShadowStyle } from "./ShadowStyles";
import { FlexBox } from "./FlexBox";

export { withLayoutStyles, withShadowStyles, createShadowStyle, FlexBox };
