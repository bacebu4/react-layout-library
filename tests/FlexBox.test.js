import React from "react";
import renderer from "react-test-renderer";

import { FlexBox } from "../src";

describe("FlexBox", () => {
  it("renders all necessary props", () => {
    const tree = renderer
      .create(
        <FlexBox
          direction="column"
          jc="center"
          ai="center"
          mt={16}
          mr={16}
          mb={16}
          ml={16}
          pt={16}
          pr={16}
          pb={16}
          pl={16}
          w="100%"
          h="100vh"
          margin="0 auto"
          padding="0 auto"
        >
          Hey
        </FlexBox>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't renders not specified props", () => {
    const tree = renderer
      .create(
        <FlexBox
          direction="column"
          jc="center"
          ai="center"
          mr={16}
          mb={16}
          ml={16}
          pt={16}
          pr={16}
          pb={16}
          pl={16}
          w="100%"
          h="100vh"
          margin="0 auto"
          padding="0 auto"
        >
          Hey
        </FlexBox>
      )
      .toJSON();
    expect(tree).not.toHaveStyleRule("margin-top");
  });
});
