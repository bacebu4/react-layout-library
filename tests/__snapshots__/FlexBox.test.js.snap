// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexBox renders all necessary props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 16px;
  margin-right: 16px;
  margin-bottom: 16px;
  margin-top: 16px;
  margin: 0 auto;
  width: 100%;
  height: 100vh;
}

<div
  className="c0"
  direction="column"
>
  Hey
</div>
`;
